qweborf/man/qweborf.1
.TH "qweborf" "1" 
.SH "NAME" 
qweborf \(em Shares files using the HTTP protocol.
.SH "SYNOPSIS" 
.PP 
\fBqweborf\fR

.SH "DESCRIPTION" 
.PP 
Qweborf provides an easy to use interface to share local files using the HTTP protocol.
It can enable writing on the server, webdav and authentication.

.SH "AUTHOR" 
.PP 
This manual page was written by Salvo 'LtWorf' Tomaselli <tiposchi@tiscali.it> for 
the \fBDebian GNU/Linux\fP system (but may be used by others).  Permission is 
granted to copy, distribute and/or modify this document under 
the terms of the GNU General Public License 
version 3 or any later version published by the Free Software Foundation. 


CHANGELOG
0.16
- qweborf has a normal makefile

0.15
- Finally support https
- Support systemd. It's now easy to run several weborf instances.
- Can set certificate with qweborf
- Various bugfixes

0.14
- Use one less thread
- Use smaller thread pool
- Detaches when connections can be accepted (can now use Type=forking in systemd)
- Log some cases of bad request
- Bug: fixes issue with authentication request causing further requests on the same connection to fail
- Reduced memory footprint for webdav requests
- Bug: memory corruption error, only triggered with some gcc flags

0.13.5
- Added --gid option to change gid
- Improved handling of reserved characters in URI
- ContentType in directory listing
- Improved code to create the socket

0.13.4
- Workaround API breaking in PyQt
- Improvements in directory listing
- Convert things to string before showing them in qweborf

0.13.3
- Better desktop integration for qweborf

0.13.2
- Fix base64 decoding bug

0.13.1
- Port qweborf to Python3 and Qt5

0.13
- Shows last modified column in directory listing (Rev 419)
- Webdav module can send mimetype
- Bug: fixes denial of service in case of malformed header range (Rev 429)
- Bug: ported fix from 0.12.2
- Faster listing of directories with several hidden files
- PROPFIND now gives %-escaped URIs (Rev 444)
- Improved handling of Connection header in responses
- Sends reason phrase in HTTP response
- Directory listing can be cached on disk
- Can support large files
- Added GiB measure in directory listing
- Ported fix from 0.12.3 (Rev. 465)
- Reduced number of params for functions
- Uses pthread calls for TLS instead of relying on __thread that doesn't work on OsX
- Dav's PROPFIND with different required properties can be cached on disk
- Can send Content-Type
- Can be used with inetd/xinetd
- Fixes crash on invalid request (Rev 492)
- Will no longer terminate on failed accept
- Heavy code refactory
- Webdav method MOVE is now faster on directories
- Support for embedded authentication policy
- Support for If-Range header (Rev 516)
- Migrated to GNU autotools
- Now logs HTTP status code for each request
- Ships qweborf: QT GUI for simple file sharing
- Truncates files on PUT
- Can send directories as tar.gz files

0.12.5
- Fixes DoS occurring with malformed fields in HTTP request

0.12.4
- Fixes DoS occurring on malformed HTTP request, backported from trunk

0.12.3
- Fixes directory traversal vulnerability

0.12.2
- Fixes DoS occurring when wide chars are used in some headers

0.12.1
- Fixes DoS occurring on malformed Range header, backported from trunk

0.12
- Fixed bug: now threads are closed again (Rev 305)
- Using keep-alive with CGI only in case of small pages (Rev 306)
- Reduced maximum size of generated pages (Rev 307)
- Adjusted size of the buffers (Rev 307)
- ETag headers are now sent only if the resource is static (Rev 308)
- "Moved permanently" code is returned instead of "see other" when client requests directory without ending / (Rev 209)
- Added wrapper for CGI python, which replaces the previous one (Rev 312)
- Support for custom CGI formats
- LSB tags added to the init script (Rev 323)
- Support to PUT,DELETE,OPTIONS methods
- Fixed bug: some file descriptors caused thread's termination (Rev 341)
- Added wrapper for executables CGIs
- Stub for webdav
- Implemented MKCOL, COPY, MOVE methods
- Bug: socket were closed randomly on non GET/POST methods
- Bug: tamed buffers do not affect subsequent connections on the same thread
- Daemon is able to start authentication server too, reading about it from weborf.conf (Rev 359)
- Better and more resistant xml parsing of PROPFIND requests (Rev 363)
- Bug: addresses are properly converted into strings (Rev 373)
- CGI does chdir into script's directory (Rev 374)
- Enables logging
- Uses reentrant version of readdir (Rev 381)
- Added more controls on memory allocation
- Bug: buffer size corrected in queue.c
- Logs insufficient memory
- Can compile on OpenSolaris (manual edit of Makefile is needed)
- Edited instance.h to workarond OpenSolaris bug #6775831
- Can use SIGUSR1 to print internal status on stdout
- Edited daemon script to pass the new lintian's test
- Attempt to resolve race-condition by refactoring
- Changed astyle parameters to use the new version
- Sorts content of directories when generating html index
- Bug: replaced unsafe sprintf with snprintf (Rev 415)

0.11
- Support for ETag header
- Support for If-None-Match header. Weborf will not re-send files that are cached into the browser
- Cleaned up some send_header functions
- Fixed bug: malformed range header was able to cause segfault (Rev 251)
- Support for Last-Modified header (Rev 252)
- Introduced optimizations in parsing request and sending an header
- Code refactory to reduce amount of passed parameters
- Will now send connection header if an old client is using keep-alive
- Corrected sending post data to cgi-bin
- Fixed REQUEST_URI header which was broken some revisions ago
- Doesn't send CGI errors and warning to the client anymore
- Shows CGI errors on standard error, but an option in options.h can hide them
- Optimizations in CGI
- Reduced system calls number
- Reduced number of strlen and strstr
- Fixed warnings in manpages

0.10
- Pipe with scripts are now closed
- Support for cgi-php
- Fixed some possibilities of buffer overflow
- Enlarged default head's buffer, necessary for script generated headers
- When the script is terminated for timeout, error 500 is sent to the client
- Fixed error in cgi wrapper that didn't allow to use \r\n\r\n sequence in output
- Compile on MacOsX without needing to edit the code
- Embedded support for cgi
- Support for list of index files
- List of index files changable in weborf.conf
- Refactory of code to send/execute pages and generate dir listing
- Changed default optimizations
- Support for virtualhost
- Support for virtualhost in weborf.conf
- Support for gzip encoding, when supported by client

qweborf/qweborf.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
# Weborf
# Copyright (C) 2010  Salvo "LtWorf" Tomaselli
#
# Weborf is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.
#
# author Salvo "LtWorf" Tomaselli <tiposchi@tiscali.it>

from PyQt5 import QtGui, QtWidgets, QtCore
import sys
import os

import qweborf.main as main
import qweborf.whelper as whelper
import qweborf.nhelper as nhelper


class qweborfForm (QtWidgets.QWidget):

    DBG_DEFAULT = 0
    DBG_WARNING = 1
    DBG_ERROR = 2
    DBG_NOTICE = 3

    def setUi(self, ui) -> None:
        self.ui = ui
        self.weborf = whelper.weborf_runner(self)
        self.started = False

        if self.weborf.version >= '0.13':
            self.ui.chkTar.setEnabled(True)
        else:
            self.ui.chkTar.setEnabled(False)

        if not self.weborf.webdav:
            self.ui.chkDav.setEnabled(False)
            self.ui.chkWrite.setEnabled(False)

        if not self.weborf.https:
            self.ui.txtKey.setEnabled(False)
            self.ui.txtCert.setEnabled(False)
            self.ui.cmdOpenCert.setEnabled(False)
            self.ui.cmdOpenKey.setEnabled(False)

        # Listing addresses
        for i in nhelper.getaddrs(self.weborf.ipv6):
            self.ui.cmbAddress.addItem(i, None)

        self.ui.chkNAT.setEnabled(nhelper.can_redirect())

        self.defaultdir = QtCore.QStandardPaths.writableLocation(
            QtCore.QStandardPaths.HomeLocation)

        initialdir = self.defaultdir

        if len(sys.argv) > 1 and os.path.exists(sys.argv[1]):
            initialdir = sys.argv[1]

        self.ui.txtPath.setText(initialdir)

    def logger(self, data, level=DBG_DEFAULT):
        '''logs an entry, showing it in the GUI'''
        if level == self.DBG_WARNING:
            data = '<font color="orange"><strong>WARNING</strong></font>: %s' % data
        elif level == self.DBG_ERROR:
            data = '<font color="red"><strong>ERROR</strong></font>: %s' % data
        elif level == self.DBG_NOTICE:
            data = '<font color="green"><strong>NOTICE</strong></font>: %s' % data
        self.ui.txtLog.moveCursor(QtGui.QTextCursor.End)
        self.ui.txtLog.insertHtml(data + '<br>')
        self.ui.txtLog.moveCursor(QtGui.QTextCursor.End)

    def setDefaultValues(self):
        '''Sets default values into the form GUI. It has to be
        called after the form has been initialized'''
        pass

    def stop_sharing(self):
        if self.weborf.stop():
            self.ui.cmdStart.setEnabled(True)
            self.ui.cmdStop.setEnabled(False)
            self.ui.tabWidget.setEnabled(True)
            self.started = False

    def about(self):

        self.logger('<hr><strong>Qweborf 0.15</strong>')
        self.logger('This program comes with ABSOLUTELY NO WARRANTY.'
                    ' This is free software, and you are welcome to redistribute it under certain conditions.'
                    ' For details see the <a href="http://www.gnu.org/licenses/gpl.html">GPLv3 Licese</a>.')
        self.logger(
            '<a href="http://ltworf.github.io/weborf/">Homepage</a>')
        self.logger(
            'Salvo \'LtWorf\' Tomaselli <a href="mailto:tiposchi@tiscali.it">&lt;tiposchi@tiscali.it&gt;</a>')

    def terminate(self):
        if self.started:
            self.stop_sharing()
        sys.exit(0)

    def start_sharing(self):

        options = {}  # Dictionary with the chosen options

        options['path'] = self.ui.txtPath.text()

        if self.ui.chkEnableAuth.isChecked():
            options['username'] = self.ui.txtUsername.text()
            options['password'] = self.ui.txtPassword.text()
        else:
            options['username'] = None
            options['password'] = None

        if self.ui.txtKey.text() or self.ui.txtCert.text():
            options['key'] = self.ui.txtKey.text()
            options['cert'] = self.ui.txtCert.text()
        else:
            options['key'] = None
            options['cert'] = None

        options['port'] = self.ui.spinPort.value()

        options['dav'] = self.ui.chkDav.isChecked()
        if self.ui.chkDav.isChecked():
            options['write'] = self.ui.chkWrite.isChecked()
        else:
            options['write'] = False

        options['tar'] = self.ui.chkTar.isChecked()

        if self.ui.cmbAddress.currentIndex() == 0:
            options['ip'] = None
        else:
            options['ip'] = str(self.ui.cmbAddress.currentText())

        if self.weborf.start(options):
            self.ui.cmdStart.setEnabled(False)
            self.ui.cmdStop.setEnabled(True)
            self.ui.tabWidget.setEnabled(False)
            self.started = True

        if self.ui.chkNAT.isChecked() and self.started==True:
            QtCore.QCoreApplication.processEvents()
            self.logger('Trying to use UPnP to open a redirection in the NAT device. Please wait...')
            QtCore.QCoreApplication.processEvents()
            external_addr=nhelper.externaladdr()
            QtCore.QCoreApplication.processEvents()
            self.logger('Public IP address %s' % str(external_addr))
            QtCore.QCoreApplication.processEvents()
            if external_addr!=None:
                redirection=nhelper.open_nat(options['port'])
                if redirection is not None:
                    self.redirection=redirection
                    url='http://%s:%d/' % (external_addr,redirection.eport)
                    logentry='Public address: <a href="%s">%s</a>' % (url,url)
                    self.logger(logentry)

    def select_cert(self) -> None:
        fname = QtWidgets.QFileDialog.getOpenFileName(
            self,
            'Certificate',
            self.ui.txtCert.text(),
            filter='PEM certificates(*.pem);;All files(*)',
        )[0]
        if fname:
            self.ui.txtCert.setText(fname)


    def select_key(self) -> None:
        fname = QtWidgets.QFileDialog.getOpenFileName(
            self,
            'Certificate',
            self.ui.txtKey.text(),
        )[0]
        if fname:
            self.ui.txtKey.setText(fname)

    def select_path(self) -> None:
        current = self.ui.txtPath.text()
        if len(current) == 0:
            current = self.defaultdir
        dirname = QtWidgets.QFileDialog.getExistingDirectory(
            self,
            'Directory to share',
            current,
            QtWidgets.QFileDialog.ShowDirsOnly
        )
        if len(dirname) > 0:
            self.ui.txtPath.setText(dirname)


def q_main() -> None:
    import sys
    app = QtWidgets.QApplication(sys.argv)
    Form = qweborfForm()
    ui = main.Ui_Form()
    ui.setupUi(Form)
    Form.setUi(ui)
    Form.show()
    res = app.exec_()
    Form.terminate()
    sys.exit(res)

if __name__ == "__main__":
    q_main()
